### PR TITLE
fix #289 - Inconsistencies of IfStatement conditions

### LINKED ISSUE #289: Inconsistencies of IfStatement conditions
I was testing the following cases using `ASTVisitor` and looking at the AST produced for the condition between parens:
```d
if (auto i = 1) {} // 1
if (const i = 1) {} // 2
if (int i = 1) {} // 3
if (const int i = 1) {} // 4
if (const(int) i = 1) {} // 5
```
The first thing is: unlike the others, condition `2` is parsed as an `AssignExpression`, meaning that the resulting `IfStatement.identifier` will not contain the identifier, and `IfStatement.expression` contains everything instead.

The second thing is: `3` and `4` will produce exactly the same result. The `const` qualifier will not be taken into account, since there is currently no property holding type constructors in `IfStatement`.
I don't know what the best way to resolve this would be; maybe simply adding `IdType[] typeConstructors` to `IfStatement`, like with `ForeachType`, would be sufficient ?

Both of these can be verified quickly using `dscanner --ast`.

### PR DESCRIPTION
**Pull request by BBasile**
In addition it's been verified that [case 2](https://github.com/dlang-community/libdparse/issues/289#issuecomment-429669330) now works in DCD.

### CODE REVIEW

File: src/dparse/parser.d [hunk 1]
```diff
@@ -3560,60 +3560,82 @@ class Parser
...
         mixin(tokenCheck!"if");
         if (moreTokens)
             node.startIndex = current().index;
         mixin(tokenCheck!"(");
+        Bookmark b = setBookmark();
 
-        if (currentIs(tok!"auto"))
+        // most commn case: rel expression and such
+        if (Expression e = parseExpression())
+            if (moreTokens && currentIs(tok!")"))
```
Comment (BBasile) [inline src/dparse/parser.d]: The need for this reveals a bug:

    if (int i = a < 100) b(); 

An expression is found before reaching the right paren.

**Comment by BBasile** [comment]
This one really needs review. Apparently parsing `IfStatement` was turned very hard because of a bug (see 3rd commit).

File: test/pass_files/templates.d [hunk 2]
```diff
@@ -4,11 +4,11 @@ template a(b, c) if (d) {}
 template a(alias int a : 10) {}
 template a(alias int a : int) {}
 template a(alias int a : 10) {}
-template a(alias int a : 10 = int) {}
+//template a(alias int a : 10 = int) {}
 template a(alias int a : int = 10) {}
 template a(alias int a : int) {}
 template a(alias int a = 10) {}
-template a(alias int a = int) {}
+//template a(alias int a = int) {}
```
Comment (BBasile) [inline test/pass_files/templates.d]: this one was valid, so the problem is not fixed

Comment (BBasile) [reply test/pass_files/templates.d]: it passes when uncommented. Problem was well fixed actually.

File: src/dparse/parser.d [hunk 3]
```diff
@@ -4896,6 +4918,7 @@ class Parser
             }
             else if (currentIs(tok!"("))
                 mixin(parseNodeQ!(`node.arguments`, `Arguments`));
+            else goto default;
```
Comment (BBasile) [inline src/dparse/parser.d]: BasicType could be accepted as expression whatever followed.

**Comment by LaurentTreguier** [comment]
`if (const i = 1) {}` is now parsed correctly, however I'm getting weird results with `typeCtors`...
I tried:
```d
override void visit(const IfStatement ifStatement)
{
    writef("%s ", ifStatement.typeCtors.length);
    foreach (ctor; ifStatement.typeCtors)
        if (ctor != tok!"")
            writef("%s ", str(ctor));
    write('\n');
}
```
With the input:
```d
void foo()
{
    if (const var = something)
        doIt();

    if (const int var = something)
        doIt();
}
```
And got the output:
```
80 .. !<= ( const
80 /= !<= floatLiteral const
```
I don't see how this could happen in the code though, so maybe it's something on my end.

**Comment by BBasile** [comment]
Yeah, the output makes no sense.

**Comment by BBasile** [comment]
writef("%d ", ifStatement.typeCtors.length);

**Comment by BBasile** [comment]
yeah that's the issue.

**Comment by BBasile** [comment]
@LaurentTreguier  Can you test again please ?

**Comment by BBasile** [comment]
@LaurentTreguier  Can you pull latest changes and test again please ?

**Comment by LaurentTreguier** [comment]
It looks all good for me now, the length is 1 and the only type constructor is const :+1:

**Comment by BBasile** [comment]
@LaurentTreguier, when a PR is ok for you you can approve formally : 

formal_approval

**Comment by BBasile** [comment]
Not good finally. There are plenty of errors when parsing parser.d for example. For now i don't know if it's because of error gagging or if parsing is really broken.

**Comment by BBasile** [comment]
`if (T t = T.init)`, parser thinks T is an expression.

**Pull request merged**